### PR TITLE
feat(copilot): Auto-save binary block outputs using content-based detection

### PR DESCRIPTION
**Pull request by Otto-AGPT**
## Summary

When CoPilot executes blocks that produce binary outputs (images, PDFs), the data is now automatically detected and saved to the workspace, replacing raw base64 data with `workspace://` references.

## Problem

Blocks return binary data as raw base64 strings in the tool response. This causes:
1. **Massive token waste** — 17,000+ output tokens for a single PDF
2. **Potential data corruption** — LLM may truncate/hallucinate characters
3. **Poor UX** — Users wait while the LLM slowly types binary data

## Key Learning from Failed Attempts

**Field-name detection (PRs #11968, #11972) does not work** because:
- Code execution blocks return base64 in `stdout_logs`, NOT in fields named `png/jpeg/pdf`
- The structured `png/jpeg/pdf` fields only populate from Jupyter display mechanisms
- Real-world code: `print(base64_pdf)` → goes to `stdout_logs`

## Solution: Content-Based Detection

Scan **ALL string values** in block outputs recursively, regardless of field name:

1. **Data URI detection** — `data:image/png;base64,...` with mimetype whitelist
2. **Magic number detection** — For raw base64 > 1KB, decode and check file signatures (PNG, JPEG, PDF, GIF, WebP)

### Safeguards Against False Positives
- Size threshold > 1KB
- Regex pre-filter (pure base64 characters only)
- Strict base64 validation
- Magic number match required
- Mimetype whitelist for data URIs

Combined: false positive risk is essentially zero.

## Changes

- **New:** `binary_output_processor.py` — Content-based detection and saving (~230 lines)
- **Modified:** `run_block.py` — Integration point (~10 lines)
- **New:** `test_binary_output_processor.py` — Comprehensive tests (~400 lines)

## Testing

Unit tests cover:
- Data URI detection (PNG, PDF, with whitelist)
- Raw base64 with magic numbers (PNG, JPEG, PDF, GIF, WebP)
- Size threshold
- Deduplication via content hash
- Nested dict/list traversal
- Graceful degradation on save failure
- The exact failing case: base64 in `stdout_logs`

## Evidence

**Failed test (session `9268c491-7ae5-43fd-83fc-2ae4244dac45`):**
- Code execution generated PDF, printed base64 to stdout
- Output: `{"stdout_logs": ["...BASE64..."], "results": [[]]}`
- Field-name detection missed it; **this PR catches it**

**Expected result:** Token usage drops ~97% (17k → ~500 tokens)

Closes SECRT-1887

### CODE REVIEW

**Comment by Torantulino** [comment]
@claude please carefully review this PR.

**Comment by Torantulino** [comment]
@claude please carefully review this PR.

File: autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+logger = logging.getLogger(__name__)
+
+# Minimum decoded size to process (filters out small base64 strings)
+MIN_DECODED_SIZE = 1024  # 1KB
+
+# Pattern to find base64 chunks in text (at least 100 chars to be worth checking)
+# Matches continuous base64 characters (with optional whitespace for line wrapping),
+# optionally ending with = padding
+EMBEDDED_BASE64_PATTERN = re.compile(r"[A-Za-z0-9+/\s]{100,}={0,2}")
```
[the remark this answers is not in the document]
Comment (Otto-AGPT) [reply autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py:33]: Acknowledged as an accepted limitation. The greedy regex could theoretically combine adjacent base64 strings separated by whitespace, but this would cause a false negative (skip detection), not data corruption. For the actual use case (single base64 block in stdout_logs), this works correctly.

File: autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+import base64
+import binascii
+import hashlib
+import logging
+import re
+import uuid
+from typing import Any, Optional
+
+from backend.util.file import sanitize_filename
```
[the remark this answers is not in the document]
Comment (Otto-AGPT) [reply autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py:22]: Acknowledged as an accepted limitation. The greedy regex could theoretically combine adjacent base64 strings separated by whitespace, but this would cause a false negative (skip detection), not data corruption. For the actual use case (single base64 block in stdout_logs), this works correctly. Making the regex non-greedy or removing \\s would introduce other edge cases with line-wrapped base64.

**Comment by Torantulino** [comment]
Tested in dev: 
image

File: autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+from backend.util.workspace import WorkspaceManager
+
+logger = logging.getLogger(__name__)
+
+# Minimum decoded size to process (filters out small base64 strings)
+MIN_DECODED_SIZE = 1024  # 1KB
```
Comment (ntindle) [inline autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py:28]: we have a size for this set in read file tool i think? its one of the workspace tools. should this be globally set or each thing have its own?

File: autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+# Magic numbers for binary file detection
+MAGIC_SIGNATURES = [
+    (b"\x89PNG\r\n\x1a\n", "png"),
+    (b"\xff\xd8\xff", "jpg"),
+    (b"%PDF-", "pdf"),
+    (b"GIF87a", "gif"),
+    (b"GIF89a", "gif"),
+    (b"RIFF", "webp"),  # Also check content[8:12] == b'WEBP'
+]
```
Comment (ntindle) [inline autogpt_platform/backend/backend/copilot/tools/binary_output_processor.py:43]: some files can be double encoded with two magic signatures. be aware

**Comment by Torantulino** [comment]
@claude please carefully review this PR, use sub-agents liberally, make no assumptions, and work from first principles thinking.

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR adds a binary output processor that detects embedded base64 binary content (PDF, PNG, JPG, GIF, WebP) in block execution outputs, saves the binary data to the user's workspace, and replaces the base64 with `workspace://` references. This reduces LLM output token usage by ~97% for file generation tasks. Comprehensive test suite included.

---

## Architecture

- **`binary_output_processor.py`**: New module with the core detection and replacement logic
- **Detection**: Regex-based base64 pattern matching → decode → magic number validation → workspace save → reference replacement
- **`run_block.py`**: Integration point — creates `WorkspaceManager` per execution, passes outputs through processor
- **Content deduplication**: SHA-256 hash cache prevents saving identical binaries multiple times
- **Virus scanning**: Content is scanned before saving via `scan_content_safe`

---

## Issues

### ⚠️ Broad regex could cause performance issues on large text outputs
**Severity: warning**

The regex `[A-Za-z0-9+\/\s]{100,}={0,2}` matches any 100+ char alphanumeric string. For large text outputs (e.g., log files, code), this will generate many false candidate matches that each need base64 decoding attempts at multiple offsets (0, 4, 8, ..., 32). The 8 decode attempts per match × N matches could be expensive.

Consider:
- Adding a pre-filter based on base64 character distribution (real base64 has ~equal char frequency)
- Skipping strings that don't contain a high density of `+` and `/` characters
- Or limiting scan to specific output fields known to contain binaries (e.g., only `stdout_logs`)

### ⚠️ Offset-based decoding could produce false positives
**Severity: warning**

The `_decode_and_validate` function tries decoding at offsets 0, 4, 8, ..., 32. At offset 32, it's decoding a substring that starts 32 chars into the matched text. This means a non-base64 string could accidentally produce valid binary with a recognized magic number at some offset. While the probability is low (magic numbers are specific), the 8-attempt approach increases the surface area.

Consider logging when decoding succeeds at non-zero offsets for monitoring.

### WorkspaceManager created per block execution
**Severity: suggestion**

In `run_block.py`, a new `WorkspaceManager` is created for every block execution:
```python
workspace_manager = WorkspaceManager(
    user_id=user_id,
    workspace_id=workspace.id,
    session_id=session.session_id,
)
```

If `WorkspaceManager` initialization involves I/O (e.g., establishing sessions), this could be expensive. Consider reusing the manager across block executions within the same session.

---

## Positive Notes

- Excellent test coverage: 518 lines of tests covering detection, markers, deduplication, virus scanning, error handling, and edge cases
- Virus scanning before saving is a good security measure
- Content deduplication via SHA-256 prevents redundant storage
- Graceful degradation — if saving fails, original output is preserved
- Marker expansion handles common patterns (`---BASE64_START---`, `[BASE64]`, `PDF_BASE64_START`)
- `sanitize_filename` usage prevents path traversal

---

## Assessment: ✅ Approve (with suggestions)

Well-designed feature with strong security practices (virus scanning, magic number validation, minimum size threshold, filename sanitization) and excellent test coverage. The regex performance concern is worth monitoring but not blocking. The offset-based decoding approach for handling marker bleed-in is clever.

_This review was created by an AI assistant (OpenHands) on behalf of the Orb Cloud review agent._
